File: Wizard Battle Web API/Services/SchoolCategoryService.cs
namespace Wizard_Battle_Web_API.Services
{
	public interface ISchoolCategoryService
	{
		Task<List<StaticSchoolCategoryResponse>> GetAll();
		Task<DirectSchoolCategoryResponse> GetById(int id);
		Task<DirectSchoolCategoryResponse> Create(SchoolCategoryRequest request);
		Task<DirectSchoolCategoryResponse> Update(int id, SchoolCategoryRequest request);
		Task<DirectSchoolCategoryResponse> Delete(int id);
	}
	public class SchoolCategoryService : ISchoolCategoryService
	{
		private readonly ISchoolCategoryRepository m_schoolCategoryRepository;
		private readonly IMapper m_mapper;


		public SchoolCategoryService(ISchoolCategoryRepository schoolCategoryRespository, IMapper mapper)
		{
			m_schoolCategoryRepository = schoolCategoryRespository;
			m_mapper = mapper;
		}


		public async Task<List<StaticSchoolCategoryResponse>> GetAll()
		{
			List<SchoolCategory> schoolCategories = await m_schoolCategoryRepository.GetAll();
			if(schoolCategories != null)
			{
				return schoolCategories.Select(schoolCategory => m_mapper.Map<SchoolCategory, StaticSchoolCategoryResponse>(schoolCategory)).ToList();
			}

			return null;
		}


		public async Task<DirectSchoolCategoryResponse> GetById(int id)
		{
			SchoolCategory category = await m_schoolCategoryRepository.GetById(id);
			if(category != null)
			{
				return m_mapper.Map<DirectSchoolCategoryResponse>(category);
			}

			return null;
		}


		public async Task<DirectSchoolCategoryResponse> Create(SchoolCategoryRequest request)
		{
			SchoolCategory category = await m_schoolCategoryRepository.Create(m_mapper.Map<SchoolCategory>(request));
			if (category != null)
			{
				return m_mapper.Map<DirectSchoolCategoryResponse>(category);
			}

			return null;
		}


		public async Task<DirectSchoolCategoryResponse> Update(int id, SchoolCategoryRequest request)
		{
			SchoolCategory category = await m_schoolCategoryRepository.Update(id, m_mapper.Map<SchoolCategory>(request));
			if (category != null)
			{
				return m_mapper.Map<DirectSchoolCategoryResponse>(category);
			}

			return null;
		}


		public async Task<DirectSchoolCategoryResponse> Delete(int id)
		{
			SchoolCategory category = await m_schoolCategoryRepository.GetById(id);
			if (category != null)
			{
				return m_mapper.Map<DirectSchoolCategoryResponse>(await m_schoolCategoryRepository.Delete(category));
			}

			return null;
		}
	}
}

File: Wizard Battle Web Client/src/app/services/school-category.service.ts
import { StaticSchoolCategoryResponse, DirectSchoolCategoryResponse, SchoolCategoryRequest } from './../_models/SchoolCategory/index';
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { environment } from 'src/environments/environment';
import { Observable } from 'rxjs';

@Injectable({
  providedIn: 'root'
})
export class SchoolCategoryService {

  constructor(private http: HttpClient) { }

  private url: string = environment.ApiUrl + "/SchoolCategory";

  public getAll(): Observable<StaticSchoolCategoryResponse[]> {
    return this.http.get<StaticSchoolCategoryResponse[]>(this.url);
  }

  public getById(id: number): Observable<DirectSchoolCategoryResponse> {
    return this.http.get<DirectSchoolCategoryResponse>(`${this.url}/${id}`);
  }

  public create(request: SchoolCategoryRequest): Observable<DirectSchoolCategoryResponse> {
    return this.http.post<DirectSchoolCategoryResponse>(this.url, request);
  }

  public update(id: number, request: SchoolCategoryRequest): Observable<DirectSchoolCategoryResponse> {
    return this.http.put<DirectSchoolCategoryResponse>(`${this.url}/${id}`, request);
  }

  public delete(id: number): Observable<DirectSchoolCategoryResponse> {
    return this.http.delete<DirectSchoolCategoryResponse>(`${this.url}/${id}`);
  }
}
